File: packages/modules/ImsMedia/tests/native/service/src/com/android/telephony/imsmedia/lib/libimsmedia/config/CallQualityTest.cpp
/**
 * Copyright (C) 2022 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include <AudioConfig.h>
#include <CallQuality.h>
#include <gtest/gtest.h>

using namespace android::telephony::imsmedia;

const int32_t kDownlinkCallQualityLevel = 0;
const int32_t kUplinkCallQualityLevel = 0;
const int32_t kCallDuration = 30000;
const int32_t kNumRtpPacketsTransmitted = 30 * 50;
const int32_t kNumRtpPacketsReceived = 30 * 50;
const int32_t kNumRtpPacketsTransmittedLost = 1;
const int32_t kNumRtpPacketsNotReceived = 1;
const int32_t kAverageRelativeJitter = 1;
const int32_t kMaxRelativeJitter = 5;
const int32_t kAverageRoundTripTime = 100;
const int32_t kCodecType = AudioConfig::CODEC_AMR_WB;
const bool kRtpInactivityDetected = false;
const bool kRxSilenceDetected = false;
const bool kTxSilenceDetected = false;
const int32_t kNumVoiceFrames = 30 * 50;
const int32_t kNumNoDataFrames = 0;
const int32_t kNumDroppedRtpPackets = 0;
const int64_t kMinPlayoutDelayMillis = 100;
const int64_t kMaxPlayoutDelayMillis = 180;
const int32_t kNumRtpSidPacketsReceived = 10;
const int32_t kNumRtpDuplicatePackets = 1;

class CallQualityTest : public ::testing::Test
{
public:
    CallQuality quality1;
    CallQuality quality2;
    CallQuality quality3;

protected:
    virtual void SetUp() override
    {
        quality1.setDownlinkCallQualityLevel(kDownlinkCallQualityLevel);
        quality1.setUplinkCallQualityLevel(kUplinkCallQualityLevel);
        quality1.setCallDuration(kCallDuration);
        quality1.setNumRtpPacketsTransmitted(kNumRtpPacketsTransmitted);
        quality1.setNumRtpPacketsReceived(kNumRtpPacketsReceived);
        quality1.setNumRtpPacketsTransmittedLost(kNumRtpPacketsTransmittedLost);
        quality1.setNumRtpPacketsNotReceived(kNumRtpPacketsNotReceived);
        quality1.setAverageRelativeJitter(kAverageRelativeJitter);
        quality1.setMaxRelativeJitter(kMaxRelativeJitter);
        quality1.setAverageRoundTripTime(kAverageRoundTripTime);
        quality1.setCodecType(kCodecType);
        quality1.setRtpInactivityDetected(kRtpInactivityDetected);
        quality1.setRxSilenceDetected(kRxSilenceDetected);
        quality1.setTxSilenceDetected(kTxSilenceDetected);
        quality1.setNumVoiceFrames(kNumVoiceFrames);
        quality1.setNumNoDataFrames(kNumNoDataFrames);
        quality1.setNumDroppedRtpPackets(kNumDroppedRtpPackets);
        quality1.setMinPlayoutDelayMillis(kMinPlayoutDelayMillis);
        quality1.setMaxPlayoutDelayMillis(kMaxPlayoutDelayMillis);
        quality1.setNumRtpSidPacketsReceived(kNumRtpSidPacketsReceived);
        quality1.setNumRtpDuplicatePackets(kNumRtpDuplicatePackets);
    }

    virtual void TearDown() override {}
};

TEST_F(CallQualityTest, TestGetterSetter) {}

TEST_F(CallQualityTest, TestParcel)
{
    android::Parcel parcel;
    quality1.writeToParcel(&parcel);
    parcel.setDataPosition(0);

    CallQuality testQuality;
    testQuality.readFromParcel(&parcel);
    EXPECT_EQ(testQuality, quality1);
}

TEST_F(CallQualityTest, TestAssign)
{
    CallQuality testQuality = quality1;
    EXPECT_EQ(quality1, testQuality);
}

TEST_F(CallQualityTest, TestEqual)
{
    quality2.setDownlinkCallQualityLevel(kDownlinkCallQualityLevel);
    quality2.setUplinkCallQualityLevel(kUplinkCallQualityLevel);
    quality2.setCallDuration(kCallDuration);
    quality2.setNumRtpPacketsTransmitted(kNumRtpPacketsTransmitted);
    quality2.setNumRtpPacketsReceived(kNumRtpPacketsReceived);
    quality2.setNumRtpPacketsTransmittedLost(kNumRtpPacketsTransmittedLost);
    quality2.setNumRtpPacketsNotReceived(kNumRtpPacketsNotReceived);
    quality2.setAverageRelativeJitter(kAverageRelativeJitter);
    quality2.setMaxRelativeJitter(kMaxRelativeJitter);
    quality2.setAverageRoundTripTime(kAverageRoundTripTime);
    quality2.setCodecType(kCodecType);
    quality2.setRtpInactivityDetected(kRtpInactivityDetected);
    quality2.setRxSilenceDetected(kRxSilenceDetected);
    quality2.setTxSilenceDetected(kTxSilenceDetected);
    quality2.setNumVoiceFrames(kNumVoiceFrames);
    quality2.setNumNoDataFrames(kNumNoDataFrames);
    quality2.setNumDroppedRtpPackets(kNumDroppedRtpPackets);
    quality2.setMinPlayoutDelayMillis(kMinPlayoutDelayMillis);
    quality2.setMaxPlayoutDelayMillis(kMaxPlayoutDelayMillis);
    quality2.setNumRtpSidPacketsReceived(kNumRtpSidPacketsReceived);
    quality2.setNumRtpDuplicatePackets(kNumRtpDuplicatePackets);
    EXPECT_EQ(quality2, quality1);
}

TEST_F(CallQualityTest, TestNotEqual)
{
    quality2.setDownlinkCallQualityLevel(kDownlinkCallQualityLevel);
    quality2.setUplinkCallQualityLevel(kUplinkCallQualityLevel);
    quality2.setCallDuration(kCallDuration);
    quality2.setNumRtpPacketsTransmitted(kNumRtpPacketsTransmitted);
    quality2.setNumRtpPacketsReceived(kNumRtpPacketsReceived);
    quality2.setNumRtpPacketsTransmittedLost(kNumRtpPacketsTransmittedLost);
    quality2.setNumRtpPacketsNotReceived(kNumRtpPacketsNotReceived);
    quality2.setAverageRelativeJitter(5);
    quality2.setMaxRelativeJitter(kMaxRelativeJitter);
    quality2.setAverageRoundTripTime(kAverageRoundTripTime);
    quality2.setCodecType(kCodecType);
    quality2.setRtpInactivityDetected(kRtpInactivityDetected);
    quality2.setRxSilenceDetected(kRxSilenceDetected);
    quality2.setTxSilenceDetected(kTxSilenceDetected);
    quality2.setNumVoiceFrames(kNumVoiceFrames);
    quality2.setNumNoDataFrames(kNumNoDataFrames);
    quality2.setNumDroppedRtpPackets(kNumDroppedRtpPackets);
    quality2.setMinPlayoutDelayMillis(kMinPlayoutDelayMillis);
    quality2.setMaxPlayoutDelayMillis(kMaxPlayoutDelayMillis);
    quality2.setNumRtpSidPacketsReceived(kNumRtpSidPacketsReceived);
    quality2.setNumRtpDuplicatePackets(kNumRtpDuplicatePackets);
    EXPECT_NE(quality2, quality1);

    quality3.setDownlinkCallQualityLevel(kDownlinkCallQualityLevel);
    quality3.setUplinkCallQualityLevel(kUplinkCallQualityLevel);
    quality3.setCallDuration(kCallDuration);
    quality3.setNumRtpPacketsTransmitted(kNumRtpPacketsTransmitted);
    quality3.setNumRtpPacketsReceived(kNumRtpPacketsReceived);
    quality3.setNumRtpPacketsTransmittedLost(kNumRtpPacketsTransmittedLost);
    quality3.setNumRtpPacketsNotReceived(kNumRtpPacketsNotReceived);
    quality3.setAverageRelativeJitter(kAverageRelativeJitter);
    quality3.setMaxRelativeJitter(kMaxRelativeJitter);
    quality3.setAverageRoundTripTime(kAverageRoundTripTime);
    quality3.setCodecType(AudioConfig::CODEC_AMR);
    quality3.setRtpInactivityDetected(kRtpInactivityDetected);
    quality3.setRxSilenceDetected(kRxSilenceDetected);
    quality3.setTxSilenceDetected(kTxSilenceDetected);
    quality3.setNumVoiceFrames(kNumVoiceFrames);
    quality3.setNumNoDataFrames(kNumNoDataFrames);
    quality3.setNumDroppedRtpPackets(kNumDroppedRtpPackets);
    quality3.setMinPlayoutDelayMillis(kMinPlayoutDelayMillis);
    quality3.setMaxPlayoutDelayMillis(kMaxPlayoutDelayMillis);
    quality3.setNumRtpSidPacketsReceived(kNumRtpSidPacketsReceived);
    quality3.setNumRtpDuplicatePackets(kNumRtpDuplicatePackets);
    EXPECT_NE(quality3, quality1);
}

File: packages/modules/ImsMedia/tests/unit/src/com/android/telephony/imsmedia/CallQualityTest.java
/**
 * Copyright (C) 2022 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.android.telephony.imsmedia;

import static com.google.common.truth.Truth.assertThat;

import android.os.Parcel;
import android.telephony.CallQuality;

import androidx.test.runner.AndroidJUnit4;

import org.junit.Test;
import org.junit.runner.RunWith;

@RunWith(AndroidJUnit4.class)
public class CallQualityTest {
    private static final int DOWNLINKCALLQUALITYLEVEL = 1;
    private static final int UPLINKCALLQUALITYLEVEL = 2;
    private static final int CALLDURATION = 40000;
    private static final int NUMRTPPACKETSTRANSMITTED = 1600;
    private static final int NUMRTPPACKETSRECEIVED = 1700;
    private static final int NUMRTPPACKETSTRANSMITTEDLOST = 400;
    private static final int NUMRTPPACKETSNOTRECEIVED = 42;
    private static final int AVERAGERELATIVEJITTER = 30;
    private static final int MAXRELATIVEJITTER = 40;
    private static final int AVERAGEROUNDTRIPTIME = 100;
    private static final int CODECTYPE = 1;
    private static final boolean RTPINACTIVITYDETECTED = true;
    private static final boolean RXSILENCEDETECTED = false;
    private static final boolean TXSILENCEDETECTED = false;
    private static final int NUMVOICEFRAMES = 100;
    private static final int NUMNODATAFRAMES = 20;
    private static final int NUMDROPPEDRTPPACKETS = 40;
    private static final long MINPLAYOUTDELAYMILLIS = 200;
    private static final long MAXPLAYOUTDELAYMILLIS = 1000;
    private static final int NUMRTPSIDPACKETSRECEIVED = 20;
    private static final int NUMRTPDUPLICATEPACKETS = 50;

    static CallQuality createCallQuality() {
        return new CallQuality.Builder()
                .setDownlinkCallQualityLevel(DOWNLINKCALLQUALITYLEVEL)
                .setUplinkCallQualityLevel(UPLINKCALLQUALITYLEVEL)
                .setCallDurationMillis(CALLDURATION)
                .setNumRtpPacketsTransmitted(NUMRTPPACKETSTRANSMITTED)
                .setNumRtpPacketsReceived(NUMRTPPACKETSRECEIVED)
                .setNumRtpPacketsTransmittedLost(NUMRTPPACKETSTRANSMITTEDLOST)
                .setNumRtpPacketsNotReceived(NUMRTPPACKETSNOTRECEIVED)
                .setAverageRelativeJitter(AVERAGERELATIVEJITTER)
                .setMaxRelativeJitter(MAXRELATIVEJITTER)
                .setAverageRoundTripTimeMillis(AVERAGEROUNDTRIPTIME)
                .setCodecType(CODECTYPE)
                .setRtpInactivityDetected(RTPINACTIVITYDETECTED)
                .setIncomingSilenceDetectedAtCallSetup(RXSILENCEDETECTED)
                .setOutgoingSilenceDetectedAtCallSetup(TXSILENCEDETECTED)
                .setNumVoiceFrames(NUMVOICEFRAMES)
                .setNumNoDataFrames(NUMNODATAFRAMES)
                .setNumDroppedRtpPackets(NUMDROPPEDRTPPACKETS)
                .setMinPlayoutDelayMillis(MINPLAYOUTDELAYMILLIS)
                .setMaxPlayoutDelayMillis(MAXPLAYOUTDELAYMILLIS)
                .setNumRtpSidPacketsReceived(NUMRTPSIDPACKETSRECEIVED)
                .setNumRtpDuplicatePackets(NUMRTPDUPLICATEPACKETS)
                .build();
    }

    static android.hardware.radio.ims.media.CallQuality createHalCallQuality() {
        final android.hardware.radio.ims.media.CallQuality callQuality =
                new android.hardware.radio.ims.media.CallQuality();
        callQuality.downlinkCallQualityLevel = DOWNLINKCALLQUALITYLEVEL;
        callQuality.uplinkCallQualityLevel = UPLINKCALLQUALITYLEVEL;
        callQuality.callDuration = CALLDURATION;
        callQuality.numRtpPacketsTransmitted = NUMRTPPACKETSTRANSMITTED;
        callQuality.numRtpPacketsReceived = NUMRTPPACKETSRECEIVED;
        callQuality.numRtpPacketsTransmittedLost = NUMRTPPACKETSTRANSMITTEDLOST;
        callQuality.numRtpPacketsNotReceived = NUMRTPPACKETSNOTRECEIVED;
        callQuality.averageRelativeJitter = AVERAGERELATIVEJITTER;
        callQuality.maxRelativeJitter = MAXRELATIVEJITTER;
        callQuality.averageRoundTripTime = AVERAGEROUNDTRIPTIME;
        callQuality.codecType = CODECTYPE;
        callQuality.rtpInactivityDetected = RTPINACTIVITYDETECTED;
        callQuality.rxSilenceDetected = RXSILENCEDETECTED;
        callQuality.txSilenceDetected = TXSILENCEDETECTED;
        callQuality.numVoiceFrames = NUMVOICEFRAMES;
        callQuality.numNoDataFrames = NUMNODATAFRAMES;
        callQuality.numDroppedRtpPackets = NUMDROPPEDRTPPACKETS;
        callQuality.minPlayoutDelayMillis = MINPLAYOUTDELAYMILLIS;
        callQuality.maxPlayoutDelayMillis = MAXPLAYOUTDELAYMILLIS;
        callQuality.numRtpSidPacketsReceived = NUMRTPSIDPACKETSRECEIVED;
        callQuality.numRtpDuplicatePackets = NUMRTPDUPLICATEPACKETS;
        return callQuality;
    }

    @Test
    public void testConstructorAndGetters() {
        CallQuality callQuality = createCallQuality();

        assertThat(callQuality.getDownlinkCallQualityLevel()).isEqualTo(DOWNLINKCALLQUALITYLEVEL);
        assertThat(callQuality.getUplinkCallQualityLevel()).isEqualTo(UPLINKCALLQUALITYLEVEL);
        assertThat(callQuality.getCallDuration()).isEqualTo(CALLDURATION);
        assertThat(callQuality.getNumRtpPacketsTransmitted()).isEqualTo(NUMRTPPACKETSTRANSMITTED);
        assertThat(callQuality.getNumRtpPacketsReceived()).isEqualTo(NUMRTPPACKETSRECEIVED);
        assertThat(callQuality.getNumRtpPacketsTransmittedLost())
            .isEqualTo(NUMRTPPACKETSTRANSMITTEDLOST);
        assertThat(callQuality.getNumRtpPacketsNotReceived()).isEqualTo(NUMRTPPACKETSNOTRECEIVED);
        assertThat(callQuality.getAverageRelativeJitter()).isEqualTo(AVERAGERELATIVEJITTER);
        assertThat(callQuality.getMaxRelativeJitter()).isEqualTo(MAXRELATIVEJITTER);
        assertThat(callQuality.getAverageRoundTripTime()).isEqualTo(AVERAGEROUNDTRIPTIME);
        assertThat(callQuality.getCodecType()).isEqualTo(CODECTYPE);
        assertThat(callQuality.isRtpInactivityDetected()).isEqualTo(RTPINACTIVITYDETECTED);
        assertThat(callQuality.isIncomingSilenceDetectedAtCallSetup()).isEqualTo(RXSILENCEDETECTED);
        assertThat(callQuality.isOutgoingSilenceDetectedAtCallSetup()).isEqualTo(TXSILENCEDETECTED);
        assertThat(callQuality.getNumVoiceFrames()).isEqualTo(NUMVOICEFRAMES);
        assertThat(callQuality.getNumNoDataFrames()).isEqualTo(NUMNODATAFRAMES);
        assertThat(callQuality.getNumDroppedRtpPackets()).isEqualTo(NUMDROPPEDRTPPACKETS);
        assertThat(callQuality.getMinPlayoutDelayMillis()).isEqualTo(MINPLAYOUTDELAYMILLIS);
        assertThat(callQuality.getMaxPlayoutDelayMillis()).isEqualTo(MAXPLAYOUTDELAYMILLIS);
        assertThat(callQuality.getNumRtpSidPacketsReceived()).isEqualTo(NUMRTPSIDPACKETSRECEIVED);
        assertThat(callQuality.getNumRtpDuplicatePackets()).isEqualTo(NUMRTPDUPLICATEPACKETS);
    }

    @Test
    public void testParcel() {
        CallQuality callQuality = createCallQuality();

        Parcel parcel = Parcel.obtain();
        callQuality.writeToParcel(parcel, 0);
        parcel.setDataPosition(0);

        CallQuality parcelConfig = CallQuality.CREATOR.createFromParcel(parcel);
        assertThat(callQuality).isEqualTo(parcelConfig);
    }

    @Test
    public void testEqual() {
        CallQuality callQuality1 = createCallQuality();
        CallQuality callQuality2 = createCallQuality();

        assertThat(callQuality1).isEqualTo(callQuality2);
    }

    @Test
    public void testNotEqual() {
        CallQuality callQuality1 = createCallQuality();
        CallQuality callQuality2 = new CallQuality.Builder()
                .setDownlinkCallQualityLevel(DOWNLINKCALLQUALITYLEVEL)
                .setUplinkCallQualityLevel(UPLINKCALLQUALITYLEVEL)
                .setCallDurationMillis(CALLDURATION)
                .setNumRtpPacketsTransmitted(NUMRTPPACKETSTRANSMITTED)
                .setNumRtpPacketsReceived(NUMRTPPACKETSRECEIVED)
                .setNumRtpPacketsTransmittedLost(NUMRTPPACKETSTRANSMITTEDLOST)
                .setNumRtpPacketsNotReceived((NUMRTPPACKETSNOTRECEIVED + 1))
                .setAverageRelativeJitter(AVERAGERELATIVEJITTER)
                .setMaxRelativeJitter(MAXRELATIVEJITTER)
                .setAverageRoundTripTimeMillis(AVERAGEROUNDTRIPTIME)
                .setCodecType(CODECTYPE)
                .setRtpInactivityDetected(RTPINACTIVITYDETECTED)
                .setIncomingSilenceDetectedAtCallSetup(RXSILENCEDETECTED)
                .setOutgoingSilenceDetectedAtCallSetup(TXSILENCEDETECTED)
                .setNumVoiceFrames(NUMVOICEFRAMES)
                .setNumNoDataFrames(NUMNODATAFRAMES)
                .setNumDroppedRtpPackets(NUMDROPPEDRTPPACKETS)
                .setMinPlayoutDelayMillis(MINPLAYOUTDELAYMILLIS)
                .setMaxPlayoutDelayMillis(MAXPLAYOUTDELAYMILLIS)
                .setNumRtpSidPacketsReceived(NUMRTPSIDPACKETSRECEIVED)
                .setNumRtpDuplicatePackets(NUMRTPDUPLICATEPACKETS)
                .build();
        assertThat(callQuality1).isNotEqualTo(callQuality2);
    }
}
